NeoML/src/Dnn/AutoDiffFunctions.cpp

```diff
@@ -0,0 +1,1026 @@
+		firstJacobian->GetMathEngine().AddDiagMatrixToMatrix( secondJacobian->GetData(), firstJacobian->GetData(),
+			firstJacobian->GetObjectCount(), firstJacobian->GetObjectSize(), firstJacobian->GetData() );
+		return firstJacobian;
+	}
+
+	firstJacobian->GetMathEngine().VectorAdd(firstJacobian->GetData(), secondJacobian->GetData(), firstJacobian->GetData(), firstJacobian->GetDataSize());
+	return firstJacobian;
+}
+
+CPtr<const CDnnBlob> Mult( const CDnnBlob* first, const CDnnBlob* second )
+{
+	NeoAssert( first != 0 );
+	NeoAssert( second != 0 );
+	NeoAssert( first->GetDataSize() == second->GetDataSize() );
+
+	IMathEngine& mathEngine = first->GetMathEngine();
+
+	const CTapeBlob* tapeBlob1 = dynamic_cast<const CTapeBlob*>( first );
+	IGradientTape* tape1 = tapeBlob1 != 0 ? tapeBlob1->Tape() : 0;
+	const CTapeBlob* tapeBlob2 = dynamic_cast<const CTapeBlob*>( second );
+	IGradientTape* tape2 = tapeBlob2 != 0 ? tapeBlob2->Tape() : 0;
+
+	NeoAssert( tape1 == 0 || tape2 == 0 || tape1 == tape2 );
+
+	IGradientTape* tape = tape1 != 0 ? tape1 : tape2;
+
+	CPtr<CTapeBlob> result( new CTapeBlob( tape, mathEngine, first->GetDesc() ) );
+	mathEngine.VectorEltwiseMultiply( first->GetData(), second->GetData(), result->GetData(), result->GetDataSize() );
+
+	if( tape != 0 ) {
+		CPtr<ITapeOperation> operation( new CTapeMult( *first, *second ) ); 
+		tape->Add( result, operation );
+	}
+
+	return result.Ptr();
+}
+
+CPtr<const CDnnBlob> Mult( const CDnnBlob* first, float value )
+{
+	NeoAssert( first != 0 );
+
+	CPtr<const CDnnBlob> second = Const( first->GetMathEngine(), value, first->GetDesc() );
+	return Mult( first, second );
+}
+
+CPtr<const CDnnBlob> Mult( float first, const CDnnBlob* second )
+{
+	return Mult( second, first );
+}
+
+//------------------------------------------------------------------------------------------------------------
+
+class CTapeDiv : public ITapeOperation {
+public:
+	explicit CTapeDiv( const CDnnBlob& first, const CDnnBlob& second );
+
+	CPtr<CDnnBlob> Jacobian( const CTapeBlob* var ) const override;
+
+private:
+	CPtr<const CDnnBlob> first;
+	CPtr<const CDnnBlob> second;
+};
+
+CTapeDiv::CTapeDiv( const CDnnBlob& _first, const CDnnBlob& _second ) :
+	first( &_first ),
+	second( &_second )
+{
+	NeoAssert( dynamic_cast<const CTapeBlob*>(first.Ptr()) != 0 || dynamic_cast<const CTapeBlob*>(second.Ptr()) != 0 );
+}
+
+CPtr<CDnnBlob> CTapeDiv::Jacobian( const CTapeBlob* var ) const
+{
+	CPtr<CDnnBlob> result;
+	CPtr<CDnnBlob> firstJacobian = callJacobian( first, var );
+	CPtr<CDnnBlob> secondJacobian = callJacobian( second, var );
+
+	if( firstJacobian == 0 && secondJacobian == 0 ) {
+		return 0;
+	}
+
+	IMathEngine& mathEngine = first->GetMathEngine();
+	const int gradientSize = firstJacobian != 0 ? firstJacobian->GetObjectSize() : secondJacobian->GetObjectSize();
+	const int vectorSize = first->GetDataSize();
+
+	if( firstJacobian == 0 ) {
+		if( secondJacobian->GetObjectCount() == 1 ) {

```

Comment:
math here looks wrong.

derivative here must  be `-secondGrad*first/(secind^2)`

Reply:
fixed